schema/partition.go

```diff
@@ -55,11 +57,25 @@ func (m *MetricData) PartitionID(method PartitionByMethod, partitions int32) (in
 		}
 		partition = jump.Hash(h.Sum64(), int(partitions))
 	case PartitionBySeriesWithTagsFnv:
-		h := fnv.New32a()
-		if err := writeSortedTagString(h, m.Name, m.Tags); err != nil {
-			return 0, err
+		// this was running too slow so we copied
```

Comment:
which prat of `writeSortedTagString()` was causing those allocations? was it the sort? if so, why not just skip the sort if we're sure that the tags are already sorted.

Reply:
also, can we rely on that the tags have been sorted before we get here? f.e. if we ingest the data in tsdb (which i think is going to include this code from MT, right?), are we still ensuring that the tags are sorted before hashing?

Reply:
We cannot assume that everyone is using `tsdb-gw` or `carbon-relay-ng` in their pipeline, so we have to do the same checks in all three programs.

Reply:
I investigated a bit where those additional allocations come from when you use `fnv.New32a()`. I think it is simply because the type `hash.Hash32` which gets returned from `fnv.New32a()` does not satisfy the `io.StringWriter` interface, but it does satisfy the `io.Writer` interface, so internally the call to `io.WriteString()` which happens in `writeSortedTagString()` needs to convert the given strings into byte slices before passing them to the `io.Writer`. If the `io.Writer` which gets passed into `writeSortedTagString()` would implement `io.WriteString()`, then it wouldn't do that. This means you could avoid copy-pasting the implementation of the hash function here if you'd just create your own copy of the type returned by `fnv.New32a()` , with an additional method to make it implement `io.StringWriter`.